File: TP1/Ejercicio 6/6.1/clasificadorIdealA.R
getDistance<-function(vector, mu){
	dist <- 0
	dim <- 5	
	for(i in 1:dim){
		dist <- dist + (vector[i] - mu[i]) ** 2		
	}	
	return (dist ** 0.5)
}

getClass<-function(vector, mu0,mu1){
	dist0 <- getDistance(vector, mu0)
	dist1 <- getDistance(vector, mu1)
	if (dist0 < dist1){
		return (0)
	}
	else {if (dist1 < dist0){
			return (1)
		}
		else{
			return (floor(runif(1,0,2)))
		}	
	}

}

calculateError<-function(p,mu0,mu1,v,name){
	error <- 0
	for(i in 1:9999){
		v[i] <- getClass(p[i,],mu0,mu1)
		if( v[i] != p[i,6]){
			error <- error + 1
		}
	}
	return (error / 999)
}


mu0 <- rep(-1,5)
mu1 <- rep(1,5)
v <- rep(0,99999)

p <- read.csv('/home/gabina/Facultad/Machine Learning/TP1/Ejercicio 6/6.1/gausianasA0.data')
e <- calculateError(p,mu0,mu1,v,"clasificacionIdealA1")
write(e,"data1")

p <- read.csv('/home/gabina/Facultad/Machine Learning/TP1/Ejercicio 6/6.1/gausianasA1.data')
e <- calculateError(p,mu0,mu1,v,"clasificacionIdealA2")
write(e,"data2")

p <- read.csv('/home/gabina/Facultad/Machine Learning/TP1/Ejercicio 6/6.1/gausianasA2.data')
e <- calculateError(p,mu0,mu1,v,"clasificacionIdealA3")
write(e,"data3")

p <- read.csv('/home/gabina/Facultad/Machine Learning/TP1/Ejercicio 6/6.1/gausianasA3.data')
e <- calculateError(p,mu0,mu1,v,"clasificacionIdealA4")
write(e,"data4")

p <- read.csv('/home/gabina/Facultad/Machine Learning/TP1/Ejercicio 6/6.1/gausianasA4.data')
e <- calculateError(p,mu0,mu1,v,"clasificacionIdealA5")
write(e,"data5")

File: TP1/Ejercicio 6/6.1/clasificadorIdealA.c
/* Traabajo Práctico 0 - Machine Learning - Ejercicio A - Gabina Luz Bianchi */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define CANT_PUNTOS 2000
#define DIM 4


float getDistance( float vector[DIM], float mu[DIM]){
	float dist = 0;
	int i;
	
	for ( i = 0; i < DIM; i++)
		dist = dist + pow((vector[i] - mu[i]), 2.0);
		
	return pow(dist, 0.5);
}

int getClass (float vector[DIM], float mu0[DIM],float mu1[DIM]){
	float dist0 = getDistance(vector, mu0);
	float dist1 = getDistance(vector, mu1);
	
	if (dist0 < dist1)
		return 0;
	else if (dist1 < dist0)
			return 1;
		else
			return (int) (rand() % 2);	

}



int main() {
	FILE *fp;
	FILE *readfp;
	int d, n, i, j;
	float  C, sigma;
	float *vector, *mu0, *mu1;

	d = DIM;

	srand(time(NULL));
	
	fp = fopen( "clasificacionIdealA.test", "w");
	readfp = fopen( "gausianasA.test", "r");
	
	mu0 = (float *) malloc(sizeof(float)*d);
	mu1 = (float *) malloc(sizeof(float)*d);
	vector = (float *) malloc(sizeof(float)*d);
	
		
	/* Inicializo los arreglos con los mu adecuados según el enunciado del ejercicio*/
	for (i = 0; i < d; i++){
		mu0[i] = -1;
		mu1[i] = 1;
	}


	for (i = 0; i < CANT_PUNTOS; i++){
		for ( j = 0; j < d; j++){
			fread(vector+j, 1, 1, readfp);
			printf("%f \n",vector[j]);
			fprintf(fp,"%f%c", vector[j], ',');
		}
		//fprintf(fp, "%d\n", getClass(vector, mu0, mu1));


	}

	fclose(fp);

	free(mu0);
	free(mu1);
	free(vector);


	return 0;
}
